src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentBag.cs

```diff
@@ -650,29 +483,27 @@ public bool IsEmpty
         {
             get
             {
-                if (_headList == null)
-                    return true;
-
-                bool lockTaken = false;
-                try
+                if (_headList != null)
                 {
-                    FreezeBag(ref lockTaken);
-                    ThreadLocalList currentList = _headList;
-                    while (currentList != null)
+                    bool lockTaken = false;
+                    try
                     {
-                        if (currentList._head != null)
-                        //at least this list is not empty, we return false
+                        FreezeBag(ref lockTaken);
```

Comment:
> Can this fail to take the lock?

Only in extreme cases, e.g. OOM while JIT'ing one of the helper methods.  It's primarily a holdover from running on desktop where thread aborts are possible and the more common reason for failure.